Error when generating definition from javascript (error TS9006)

# Bug Report
Hi,
our goal is to generate definition from javascript files.It will be helpful for checking types and have completion.

Our workspace environmnent is a multi project that declare project dependencies using references property in tsconfig. 

It is stopped in our case with this error TS9006: Declaration emit for this file requires using private name 'BaseClass'.

We couldn't find a satisfactory workaround. 

Can we say that it is a limitation of typescript ?

Thanks

### 🔎 Search Terms

 error TS9006: Declaration emit for this file requires using private name

Not sure if this is the same issue as #41672 

### 🕗 Version & Regression Information

No regression. It is something that we started to work on.
The issue is reproducible with typescript 4.5.5 and 4.6.0-beta.

### ⏯ Playground Link

This minimal repo illustrate our case and highlight the issue: 
https://github.com/llgcode/js-def-issue

### 💻 Code

<!-- Please post the relevant code sample here as well-->
```ts
declare module "deps/BaseClass" {
    class BaseClass {
        static extends<A>(a: A): new () => A & BaseClass;
    }
    export = BaseClass;
}
```
```ts
define("lib/ExtendedClass", ["deps/BaseClass"], 
/**
 * {typeof import("deps/BaseClass")}
 * @param  {typeof import("deps/BaseClass")} BaseClass 
 * @returns 
 */
(BaseClass) => {    
    
    const ExtendedClass = BaseClass.extends({
        f: function() {
            return "something";
        }
    });

    // Exports the module in a way tsc recognize class export 
    const module = {};
    module.exports = ExtendedClass
    return module.exports;
});

```

### 🙁 Actual behavior

```
ExtendedClass.js:1:1 - error TS9006: Declaration emit for this file requires using private name 'BaseClass' from module '"deps/BaseClass"'. An explicit type annotation may unblock declaration emit.

1 define("lib/ExtendedClass", ["deps/BaseClass"],
  ~~~~~~
```

### 🙂 Expected behavior

Generate the definition file.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -3981,8 +3981,15 @@ namespace ts {
                 return res;
             }
             const candidates = mapDefined(symbol.declarations, d => {
-                if (!isAmbientModule(d) && d.parent && hasNonGlobalAugmentationExternalModuleSymbol(d.parent)) {
-                    return getSymbolOfNode(d.parent);
+                if (!isAmbientModule(d) && d.parent){
+                    // direct children of a module
+                    if (hasNonGlobalAugmentationExternalModuleSymbol(d.parent)) {
+                        return getSymbolOfNode(d.parent);
+                    }
+                    // export ='d member of an ambient module
+                    if (isModuleBlock(d.parent) && d.parent.parent && resolveSymbol(getSymbolOfNode(d.parent.parent).exports?.get(InternalSymbolName.ExportEquals)) === symbol) {
+                        return getSymbolOfNode(d.parent.parent);
+                    }
                 }
                 if (isClassExpression(d) && isBinaryExpression(d.parent) && d.parent.operatorToken.kind === SyntaxKind.EqualsToken && isAccessExpression(d.parent.left) && isEntityNameExpression(d.parent.left.expression)) {
                     if (isModuleExportsAccessExpression(d.parent.left) || isExportsIdentifier(d.parent.left.expression)) {

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -3987,7 +3987,7 @@ namespace ts {
                         return getSymbolOfNode(d.parent);
                     }
                     // export ='d member of an ambient module
-                    if (isModuleBlock(d.parent) && d.parent.parent && resolveSymbol(getSymbolOfNode(d.parent.parent).exports?.get(InternalSymbolName.ExportEquals)) === symbol) {
+                    if (isModuleBlock(d.parent) && d.parent.parent && resolveExternalModuleSymbol(getSymbolOfNode(d.parent.parent)) === symbol) {
                         return getSymbolOfNode(d.parent.parent);
                     }
                 }

--- a/tests/baselines/reference/amdLikeInputDeclarationEmit.js
+++ b/tests/baselines/reference/amdLikeInputDeclarationEmit.js
@@ -1,0 +1,40 @@
+//// [tests/cases/compiler/amdLikeInputDeclarationEmit.ts] ////
+
+//// [typing.d.ts]
+declare function define<T=unknown>(name: string, modules: string[], ready: (...modules: unknown[]) => T);
+//// [BaseClass.d.ts]
+declare module "deps/BaseClass" {
+    class BaseClass {
+        static extends<A>(a: A): new () => A & BaseClass;
+    }
+    export = BaseClass;
+}
+//// [ExtendedClass.js]
+define("lib/ExtendedClass", ["deps/BaseClass"], 
+/**
+ * {typeof import("deps/BaseClass")}
+ * @param  {typeof import("deps/BaseClass")} BaseClass 
+ * @returns 
+ */
+(BaseClass) => {    
+    
+    const ExtendedClass = BaseClass.extends({
+        f: function() {
+            return "something";
+        }
+    });
+
+    // Exports the module in a way tsc recognize class export 
+    const module = {};
+    module.exports = ExtendedClass
+    return module.exports;
+});
+
+
+
+//// [ExtendedClass.d.ts]
+/// <reference path="../tests/cases/compiler/deps/BaseClass.d.ts" />
+export = ExtendedClass;
+declare const ExtendedClass: new () => {
+    f: () => "something";
+} & import("deps/BaseClass");

--- a/tests/baselines/reference/amdLikeInputDeclarationEmit.symbols
+++ b/tests/baselines/reference/amdLikeInputDeclarationEmit.symbols
@@ -1,0 +1,66 @@
+=== tests/cases/compiler/typing.d.ts ===
+declare function define<T=unknown>(name: string, modules: string[], ready: (...modules: unknown[]) => T);
+>define : Symbol(define, Decl(typing.d.ts, 0, 0))
+>T : Symbol(T, Decl(typing.d.ts, 0, 24))
+>name : Symbol(name, Decl(typing.d.ts, 0, 35))
+>modules : Symbol(modules, Decl(typing.d.ts, 0, 48))
+>ready : Symbol(ready, Decl(typing.d.ts, 0, 67))
+>modules : Symbol(modules, Decl(typing.d.ts, 0, 76))
+>T : Symbol(T, Decl(typing.d.ts, 0, 24))
+
+=== tests/cases/compiler/deps/BaseClass.d.ts ===
+declare module "deps/BaseClass" {
+>"deps/BaseClass" : Symbol("deps/BaseClass", Decl(BaseClass.d.ts, 0, 0))
+
+    class BaseClass {
+>BaseClass : Symbol(BaseClass, Decl(BaseClass.d.ts, 0, 33))
+
+        static extends<A>(a: A): new () => A & BaseClass;
+>extends : Symbol(BaseClass.extends, Decl(BaseClass.d.ts, 1, 21))
+>A : Symbol(A, Decl(BaseClass.d.ts, 2, 23))
+>a : Symbol(a, Decl(BaseClass.d.ts, 2, 26))
+>A : Symbol(A, Decl(BaseClass.d.ts, 2, 23))
+>A : Symbol(A, Decl(BaseClass.d.ts, 2, 23))
+>BaseClass : Symbol(BaseClass, Decl(BaseClass.d.ts, 0, 33))
+    }
+    export = BaseClass;
+>BaseClass : Symbol(BaseClass, Decl(BaseClass.d.ts, 0, 33))
+}
+=== tests/cases/compiler/ExtendedClass.js ===
+define("lib/ExtendedClass", ["deps/BaseClass"], 
+>define : Symbol(define, Decl(typing.d.ts, 0, 0))
+
+/**
+ * {typeof import("deps/BaseClass")}
+ * @param  {typeof import("deps/BaseClass")} BaseClass 
+ * @returns 
+ */
+(BaseClass) => {    
+>BaseClass : Symbol(BaseClass, Decl(ExtendedClass.js, 6, 1))
+    
+    const ExtendedClass = BaseClass.extends({
+>ExtendedClass : Symbol(ExtendedClass, Decl(ExtendedClass.js, 8, 9))
+>BaseClass.extends : Symbol(BaseClass.extends, Decl(BaseClass.d.ts, 1, 21))
+>BaseClass : Symbol(BaseClass, Decl(ExtendedClass.js, 6, 1))
+>extends : Symbol(BaseClass.extends, Decl(BaseClass.d.ts, 1, 21))
+
+        f: function() {
+>f : Symbol(f, Decl(ExtendedClass.js, 8, 45))
+
+            return "something";
+        }
+    });
+
+    // Exports the module in a way tsc recognize class export 
+    const module = {};
+>module : Symbol(module, Decl(ExtendedClass.js, 15, 9))
+
+    module.exports = ExtendedClass
+>module : Symbol(export=, Decl(ExtendedClass.js, 15, 22))
+>exports : Symbol(export=, Decl(ExtendedClass.js, 15, 22))
+>ExtendedClass : Symbol(ExtendedClass, Decl(ExtendedClass.js, 8, 9))
+
+    return module.exports;
+>module : Symbol(module, Decl(ExtendedClass.js, 15, 9))
+
+});

--- a/tests/baselines/reference/amdLikeInputDeclarationEmit.types
+++ b/tests/baselines/reference/amdLikeInputDeclarationEmit.types
@@ -1,0 +1,74 @@
+=== tests/cases/compiler/typing.d.ts ===
+declare function define<T=unknown>(name: string, modules: string[], ready: (...modules: unknown[]) => T);
+>define : <T = unknown>(name: string, modules: string[], ready: (...modules: unknown[]) => T) => any
+>name : string
+>modules : string[]
+>ready : (...modules: unknown[]) => T
+>modules : unknown[]
+
+=== tests/cases/compiler/deps/BaseClass.d.ts ===
+declare module "deps/BaseClass" {
+>"deps/BaseClass" : typeof import("deps/BaseClass")
+
+    class BaseClass {
+>BaseClass : BaseClass
+
+        static extends<A>(a: A): new () => A & BaseClass;
+>extends : <A>(a: A) => new () => A & BaseClass
+>a : A
+    }
+    export = BaseClass;
+>BaseClass : BaseClass
+}
+=== tests/cases/compiler/ExtendedClass.js ===
+define("lib/ExtendedClass", ["deps/BaseClass"], 
+>define("lib/ExtendedClass", ["deps/BaseClass"], /** * {typeof import("deps/BaseClass")} * @param  {typeof import("deps/BaseClass")} BaseClass  * @returns  */(BaseClass) => {            const ExtendedClass = BaseClass.extends({        f: function() {            return "something";        }    });    // Exports the module in a way tsc recognize class export     const module = {};    module.exports = ExtendedClass    return module.exports;}) : any
+>define : <T = unknown>(name: string, modules: string[], ready: (...modules: unknown[]) => T) => any
+>"lib/ExtendedClass" : "lib/ExtendedClass"
+>["deps/BaseClass"] : string[]
+>"deps/BaseClass" : "deps/BaseClass"
+
+/**
+ * {typeof import("deps/BaseClass")}
+ * @param  {typeof import("deps/BaseClass")} BaseClass 
+ * @returns 
+ */
+(BaseClass) => {    
+>(BaseClass) => {            const ExtendedClass = BaseClass.extends({        f: function() {            return "something";        }    });    // Exports the module in a way tsc recognize class export     const module = {};    module.exports = ExtendedClass    return module.exports;} : (BaseClass: typeof import("deps/BaseClass")) => any
+>BaseClass : typeof import("deps/BaseClass")
+    
+    const ExtendedClass = BaseClass.extends({
+>ExtendedClass : new () => { f: () => "something"; } & import("deps/BaseClass")
+>BaseClass.extends({        f: function() {            return "something";        }    }) : new () => { f: () => "something"; } & import("deps/BaseClass")
+>BaseClass.extends : <A>(a: A) => new () => A & import("deps/BaseClass")
+>BaseClass : typeof import("deps/BaseClass")
+>extends : <A>(a: A) => new () => A & import("deps/BaseClass")
+>{        f: function() {            return "something";        }    } : { f: () => "something"; }
+
+        f: function() {
+>f : () => "something"
+>function() {            return "something";        } : () => "something"
+
+            return "something";
+>"something" : "something"
+        }
+    });
+
+    // Exports the module in a way tsc recognize class export 
+    const module = {};
+>module : {}
+>{} : {}
+
+    module.exports = ExtendedClass
+>module.exports = ExtendedClass : any
+>module.exports : any
+>module : {}
+>exports : any
+>ExtendedClass : new () => { f: () => "something"; } & import("deps/BaseClass")
+
+    return module.exports;
+>module.exports : any
+>module : {}
+>exports : any
+
+});

--- a/tests/baselines/reference/importDeclFromTypeNodeInJsSource.types
+++ b/tests/baselines/reference/importDeclFromTypeNodeInJsSource.types
@@ -5,7 +5,7 @@ declare module "events" {
 >"events" : typeof import("events")
 
     namespace EventEmitter {
->EventEmitter : typeof EventEmitter
+>EventEmitter : typeof import("events")
 
         class EventEmitter {
 >EventEmitter : EventEmitter

--- a/tests/baselines/reference/importTypeAmbient.types
+++ b/tests/baselines/reference/importTypeAmbient.types
@@ -13,7 +13,7 @@ declare module "foo" {
 >Point : Point
 }
 const x: import("foo") = { x: 0, y: 0 };
->x : Point
+>x : import("foo")
 >{ x: 0, y: 0 } : { x: number; y: number; }
 >x : number
 >0 : 0

--- a/tests/baselines/reference/incompatibleExports1.types
+++ b/tests/baselines/reference/incompatibleExports1.types
@@ -23,7 +23,7 @@ declare module "baz" {
     }
  
     module c {
->c : typeof c
+>c : typeof import("baz")
 
         export var c: string;
 >c : string

--- a/tests/cases/compiler/amdLikeInputDeclarationEmit.ts
+++ b/tests/cases/compiler/amdLikeInputDeclarationEmit.ts
@@ -1,0 +1,34 @@
+// @declaration: true
+// @declarationDir: definitions
+// @emitDeclarationOnly: true
+// @checkJs: true
+// @allowJs: true
+// @filename: typing.d.ts
+declare function define<T=unknown>(name: string, modules: string[], ready: (...modules: unknown[]) => T);
+// @filename: deps/BaseClass.d.ts
+declare module "deps/BaseClass" {
+    class BaseClass {
+        static extends<A>(a: A): new () => A & BaseClass;
+    }
+    export = BaseClass;
+}
+// @filename: ExtendedClass.js
+define("lib/ExtendedClass", ["deps/BaseClass"], 
+/**
+ * {typeof import("deps/BaseClass")}
+ * @param  {typeof import("deps/BaseClass")} BaseClass 
+ * @returns 
+ */
+(BaseClass) => {    
+    
+    const ExtendedClass = BaseClass.extends({
+        f: function() {
+            return "something";
+        }
+    });
+
+    // Exports the module in a way tsc recognize class export 
+    const module = {};
+    module.exports = ExtendedClass
+    return module.exports;
+});

